src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs

```diff
@@ -249,11 +249,14 @@ public bool AreEqual(object x, object y, ref Tolerance tolerance)
 
         private static MethodInfo FirstImplementsIEquatableOfSecond(Type first, Type second)
         {
+            var pair = new KeyValuePair<Type, MethodInfo>();
+
             foreach (var xEquatableArgument in GetEquatableGenericArguments(first))
                 if (xEquatableArgument.Key.IsAssignableFrom(second))
-                    return xEquatableArgument.Value;
+                    if (pair.Key == null || pair.Key.IsAssignableFrom(xEquatableArgument.Key))
```

Comment:
I assume that you checked that GetEquatableGenericArguments always returns in the order base -> derived? Even if so, there is no guarantee that the Generic argument is more specific than that of the base. I agree that it is likely the case, so this is probably good enough.